src/helpers.js

```diff
@@ -195,6 +204,230 @@ module.exports = ({ types: t }) => {
+        // }
+        blockStatements.push(
+          t.ifStatement(
+            t.logicalExpression(
+              '||',
+              t.binaryExpression('===', depsIdentifier, t.nullLiteral()),
+              t.logicalExpression(
+                '||',
+                t.binaryExpression(
+                  '!==',
+                  t.unaryExpression('typeof', depsIdentifier),
+                  t.stringLiteral('object')
+                ),
+                t.callExpression(t.identifier('isNaN'), [
+                  t.memberExpression(depsIdentifier, t.identifier('length')),
+                ])
+              )
+            ),
+            t.blockStatement([
+              t.returnStatement(t.callExpression(t.identifier(REQUIRE), [depsIdentifier])),
+            ])
+          )
+        );
+      }
+      // Convert array-like objects to real arrays
+      // amdDeps = [].slice.call(amdDeps);
+      blockStatements.push(
+        t.expressionStatement(
+          t.assignmentExpression(
+            '=',
+            depsIdentifier,
+            t.callExpression(
+              t.memberExpression(
+                t.memberExpression(t.arrayExpression(), t.identifier('slice')),
+                t.identifier('call')
+              ),
+              [depsIdentifier]
+            )
+          )
+        )
+      );
+    }
+
+    // If we don't know that the factory is a function, then we need to check the type
+    // at runtime.
+    if (!isFunctionExpression(factory)) {
+      // define:
+      // if (typeof maybeFunction !== 'function') {
+      //   var amdFactoryResult = maybeFunction;
+      //   maybeFunction = function () {return amFactorydResult};
+      // }
+      // require:
+      // if (typeof maybeFunction !== 'function') {
+      //   maybeFunction = function () {};
+      // }
+      const resultIdentifier = getUniqueIdentifier(path.scope, AMD_FACTORY_RESULT);
+      blockStatements.push(
+        t.ifStatement(
+          t.binaryExpression(
+            '!==',
+            t.unaryExpression('typeof', factoryIdentifier),
+            t.stringLiteral('function')
+          ),
+          t.blockStatement([
+            ...(isDefineCall
+              ? [
+                  t.variableDeclaration('var', [
+                    t.variableDeclarator(resultIdentifier, factoryIdentifier),
+                  ]),
+                ]
+              : []),
+            t.expressionStatement(
+              t.assignmentExpression(
+                '=',
+                factoryIdentifier,
+                t.functionExpression(
+                  null,
+                  [],
+                  t.blockStatement([...(isDefineCall ? [t.returnStatement(resultIdentifier)] : [])])
+                )
+              )
+            ),
+          ])
+        )
+      );
+    }
+    // Invoke the factory function.
+    const factoryInvocation = t.callExpression(
+      t.memberExpression(factoryIdentifier, t.identifier('apply')),
```

Comment:
`.apply` is used so that we can pass the `this` reference.  We could use an arrow function instead, but I didn't think that arrow functions should be used in generated code.  It's necessary to preserve the `this` context because of the following common use pattern:

```javascript
require(['foo'], function(foo) {
  foo.saySomething(this.sayWhat);
}.bind(this));
```

If we use a normal (non-arrow) function call, then the `this` reference within the function will be the `window` object.

Reply:
But we're passing undefined (void 0) as the this arg every time, so isn't that just the same as calling the function without using apply?